tests/core/engine_adapter/test_mssql.py

```diff
@@ -0,0 +1,546 @@
+    target_table_name = "target_table"
+
+    def table_columns(table_name: str) -> t.Dict[str, exp.DataType]:
+        if table_name == current_table_name:
+            return {k: exp.DataType.build(v) for k, v in current_table.items()}
+        else:
+            return {k: exp.DataType.build(v) for k, v in target_table.items()}
+
+    adapter.columns = table_columns
+
+    adapter.alter_table(
+        current_table_name,
+        target_table_name,
+    )
+
+    adapter.cursor.begin.assert_called_once()
+    adapter.cursor.commit.assert_called_once()
+    adapter.cursor.execute.assert_has_calls([call(x) for x in expected])
+
+
+def test_merge_pandas(make_mocked_engine_adapter: t.Callable):
+    adapter = make_mocked_engine_adapter(MsSqlEngineAdapter)
+
+    df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
+    adapter.merge(
+        target_table="target",
+        source_table=df,
+        columns_to_types={
+            "id": exp.DataType.Type.INT,
+            "ts": exp.DataType.Type.TIMESTAMP,
+            "val": exp.DataType.Type.INT,
+        },
+        unique_key=["id"],
+    )
+    adapter.cursor.execute.assert_called_once_with(
+        'MERGE INTO "target" AS "__MERGE_TARGET__" USING (SELECT CAST("id" AS INTEGER) AS "id", CAST("ts" AS DATETIME2) AS "ts", CAST("val" AS INTEGER) AS "val" FROM (VALUES (1, 4), (2, 5), (3, 6)) AS "t"("id", "ts", "val")) AS "__MERGE_SOURCE__" ON "__MERGE_TARGET__"."id" = "__MERGE_SOURCE__"."id" '
+        'WHEN MATCHED THEN UPDATE SET "__MERGE_TARGET__"."id" = "__MERGE_SOURCE__"."id", "__MERGE_TARGET__"."ts" = "__MERGE_SOURCE__"."ts", "__MERGE_TARGET__"."val" = "__MERGE_SOURCE__"."val" '
+        'WHEN NOT MATCHED THEN INSERT ("id", "ts", "val") VALUES ("__MERGE_SOURCE__"."id", "__MERGE_SOURCE__"."ts", "__MERGE_SOURCE__"."val")'
+    )
+
+    adapter.cursor.reset_mock()
+    adapter.merge(
+        target_table="target",
+        source_table=df,
+        columns_to_types={
+            "id": exp.DataType.Type.INT,
+            "ts": exp.DataType.Type.TIMESTAMP,
+            "val": exp.DataType.Type.INT,
+        },
+        unique_key=["id", "ts"],
+    )
+    adapter.cursor.execute.assert_called_once_with(
+        'MERGE INTO "target" AS "__MERGE_TARGET__" USING (SELECT CAST("id" AS INTEGER) AS "id", CAST("ts" AS DATETIME2) AS "ts", CAST("val" AS INTEGER) AS "val" FROM (VALUES (1, 4), (2, 5), (3, 6)) AS "t"("id", "ts", "val")) AS "__MERGE_SOURCE__" ON "__MERGE_TARGET__"."id" = "__MERGE_SOURCE__"."id" AND "__MERGE_TARGET__"."ts" = "__MERGE_SOURCE__"."ts" '
+        'WHEN MATCHED THEN UPDATE SET "__MERGE_TARGET__"."id" = "__MERGE_SOURCE__"."id", "__MERGE_TARGET__"."ts" = "__MERGE_SOURCE__"."ts", "__MERGE_TARGET__"."val" = "__MERGE_SOURCE__"."val" '
+        'WHEN NOT MATCHED THEN INSERT ("id", "ts", "val") VALUES ("__MERGE_SOURCE__"."id", "__MERGE_SOURCE__"."ts", "__MERGE_SOURCE__"."val")'
+    )
+
+
+def test_replace_query_pandas(make_mocked_engine_adapter: t.Callable):
```

Comment:
I think we also need a test for `test_replace_query` with a query rather than a pandas dataframe. I expect it to yield a statement unsupported by MSSQL.

Reply:
This appears to be the same situation as above. I had noticed that each engine adapter only had tests that would expect different results than the base engine. So I copied all the tests over from the base and only kept tests in that failed and I had to change something for.

Reply:
Yeah, the shortcoming of existing tests is that they don't test against an actual database and only test against sql string expectations. This protects code from regression but requires the developer to exercise their judgement when it comes to what is and what is not supported by the target database. 

The reason why I highlighted `replace_query` and `create_index` specifically is that I suspect that their versions inherited from the base class won't produce queries compatible with MSSQL. If it turns out not to be the case, and sqlglot handles conversion correctly, then please feel free to disregard my request for adding additional tests.

Reply:
You're correct on your hunch that sqlglot is producing queries that aren't compatible with MSSQL. I'm planning to take a look to see if I can fix that. Am I correct in understanding that this would be a tsql dialect change in the sqlglot project?

Reply:
> Am I correct in understanding that this would be a tsql dialect change in the sqlglot project?

Yep, very likely. If you need any help taking on that task feel free to reach out on slack.

Reply:
Thank you! This one looked over my head so I created an issue [here](https://github.com/tobymao/sqlglot/issues/2065). I'll be interested to see how change change works in case we run into edge cases in the future!

Reply:
FYI @deschman I addressed the schema creation issue in https://github.com/tobymao/sqlglot/pull/2083, but haven't fixed the "Replacing data" issues upstream: see https://github.com/tobymao/sqlglot/issues/2065#issuecomment-1682221286.

The fixes to these issues should probably go in SQL Server's engine adapter. Based on the expected results you showed for the `INSERT INTO` statement, I think you just need to pop the `REPLACE WHERE` off of its AST and then create and prepend a `DELETE FROM` statement based on it so that it gets executed first.

Does this make sense?